lookout/style/typos/preprocessing/create_typos.py

```diff
@@ -0,0 +1,155 @@
+                         add_typo_probability: float) -> pandas.DataFrame:
+    """
+    Create artificial typos in tokens from a dataframe.
+
+    Augment some of identifiers from dataframe with TYPO_PROBABILITY,
+    consequent typos in the same word happen with ADD_TYPO_PROBABILITY each.
+    Operations happens inplace.
+    """
+    typoed_tokens = list(data[COLUMNS["TOKEN"]])
+    if COLUMNS["SPLIT"] in data.columns:
+        typoed_token_split = list(data[COLUMNS["SPLIT"]])
+
+    for row_number in tqdm(range(len(data))):
+        if typoed_tokens[row_number] is not None:
+            item = typoed_tokens[row_number]
+            if len(item) > 1 and rand_bool(typo_probability):
+                item = ""
+                while len(item) < 2:
+                    item = typoed_tokens[row_number]
+                    item = rand_typo(str(item))
+                    while rand_bool(add_typo_probability):
+                        item = rand_typo(item)
+
+            if COLUMNS["SPLIT"] in data.columns:
+                split = str(typoed_token_split[row_number]).split()
+                index = split.index(typoed_tokens[row_number])
+                split[index] = item
+                typoed_token_split[row_number] = " ".join(split)
+
+            typoed_tokens[row_number] = item
+
+    data.loc[:, COLUMNS["CORRECT_TOKEN"]] = data[COLUMNS["TOKEN"]]
+    data.loc[:, COLUMNS["TOKEN"]] = typoed_tokens
+    if COLUMNS["SPLIT"] in data.columns:
+        data.loc[:, COLUMNS["CORRECT_SPLIT"]] = data[COLUMNS["SPLIT"]]
+        data.loc[:, COLUMNS["SPLIT"]] = typoed_token_split
+    return data
+
+
+def corrupt_tokens_in_file(data_file: str, typo_probability: float, add_typo_probability: float,
+                           out_file: str, repeats: int = 1) -> None:
+    """Corrupt tokens in text file."""
+    with open(data_file, "r") as f:
+        with open(out_file, "w") as out:
+            for _ in range(repeats):
+                for line in f:
+                    tokens = []
+                    for token in line.split():
+                        item = token
+                        if rand_bool(typo_probability):
+                            item = ""
+                            while len(item) == 0:
+                                item = token
+                                item = rand_typo(str(item))
+                                while rand_bool(add_typo_probability):
+                                    item = rand_typo(item)
+                        tokens.append(item)
+                    print(" ".join(tokens), file=out)
+
+
+def train_test_split(data: pandas.DataFrame, test_portion: float,
+                     ) -> Tuple[pandas.DataFrame, pandas.DataFrame]:
+    """
+    Randomly split data on train and test.
+    """
+    test = set(random.sample(range(len(data)), int(data.shape[0] * test_portion)))
+    test_mask = [i in test for i in range(len(data))]
+    train_mask = [i not in test for i in range(len(data))]
```

Comment:
You can do random shuffle (numpy) over the indices and initialize the masks as `shuffled[:size]` and `shuffled[size:]`